### PR TITLE
Changing the handling of client claims to use JSON

### PR DESCRIPTION
**Pull request by trwalke**
Fixes #
https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/4868

**Changes proposed in this request**

**Testing**

**Performance impact**

**Documentation**
- [ ] All relevant documentation is updated.

### CODE REVIEW

**Comment by trwalke** [comment]
@localden @bgavrilMS I switched the parsing for custom claims back to the JSON parsing. Here is the change in parsing JSON strings 

The input KVP is `{ "customClaims", "{\"xms_az_claim\": [\"GUID\", \"GUID2\", \"GUID3\"]}" }`

Current: `"customClaims":"{"xms_az_claim": ["GUID", "GUID2", "GUID3"]}"`
New (fix): `"customClaims":"{\"xms_az_claim\": [\"GUID\", \"GUID2\", \"GUID3\"]}"`

Is this what we are looking for? I want to be sure this is exactly what we need to send to ESTS.

File: src/client/Microsoft.Identity.Client/Internal/JsonWebToken.cs [hunk 1]
```diff
@@ -65,18 +70,17 @@ private string CreateJsonPayload()
                 payload.Append('{');
             }
 
-            int i = 0;
-            foreach (var kvp in _claimsToSign)
+            var json = new JObject();
+
+            foreach (var claim in _claimsToSign)
             {
-                payload.Append($"\"{kvp.Key}\":\"{kvp.Value}\"");
```
Comment (trwalke) [inline src/client/Microsoft.Identity.Client/Internal/JsonWebToken.cs:71]: What was the reason for switching this to manual string parsing? Performance right? Since we are no longer parsing the default claims with JSON, but only the custom claims, I would imagine this approach will keep some of the performance gains depending on how many claims are added by the client app.

The reason I went with this approach is the JSON api will handle all of the edge cases we may miss trying to parse manually.

Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/Internal/JsonWebToken.cs:71]: SOunds good, thanks Travis. Just add some unit tests for this case, they are clearly missing.

File: tests/Microsoft.Identity.Test.Unit/PublicApiTests/ClientCredentialWithCertTest.cs [hunk 2]
```diff
@@ -506,6 +506,8 @@ public void ClientAssertionTests(bool sendX5C, bool useSha2AndPss, bool addExtra
             {
                 Assert.AreEqual("Val1", decodedToken.Payload["Key1"]);
                 Assert.AreEqual("Val2", decodedToken.Payload["Key2"]);
+                //Ensure JSON formatting is preserved
+                Assert.AreEqual("{\"xms_az_claim\": [\"GUID\", \"GUID2\", \"GUID3\"]}", decodedToken.Payload["customClaims"]);
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Unit/PublicApiTests/ClientCredentialWithCertTest.cs:510]: LGTM, anything else to add ? The PR is still a draft.

**Review by bgavrilMS — APPROVED** [review]

**Review by localden — APPROVED** [review]

File: tests/Microsoft.Identity.Test.Unit/PublicApiTests/ClientCredentialWithCertTest.cs (hunk 2 above)
Comment (trwalke) [reply tests/Microsoft.Identity.Test.Unit/PublicApiTests/ClientCredentialWithCertTest.cs:510]: Nope, I marked as ready

**Comment by bgavrilMS** [comment]
> @localden @bgavrilMS I switched the parsing for custom claims back to the JSON parsing. Here is the change in parsing JSON strings
> 
> The input KVP is `{ "customClaims", "{\"xms_az_claim\": [\"GUID\", \"GUID2\", \"GUID3\"]}" }`
> 
> Current: `"customClaims":"{"xms_az_claim": ["GUID", "GUID2", "GUID3"]}"` New (fix): `"customClaims":"{\"xms_az_claim\": [\"GUID\", \"GUID2\", \"GUID3\"]}"`
> 
> Is this what we are looking for? I want to be sure this is exactly what we need to send to ESTS.

Please see the behavior on an older MSAL to confirm.

**Comment by rayluo** [comment]
> > @localden @bgavrilMS I switched the parsing for custom claims back to the JSON parsing. Here is the change in parsing JSON strings
> > The input KVP is `{ "customClaims", "{\"xms_az_claim\": [\"GUID\", \"GUID2\", \"GUID3\"]}" }`
> > Current: `"customClaims":"{"xms_az_claim": ["GUID", "GUID2", "GUID3"]}"`
> > New (fix): `"customClaims":"{\"xms_az_claim\": [\"GUID\", \"GUID2\", \"GUID3\"]}"`
> > Is this what we are looking for? I want to be sure this is exactly what we need to send to ESTS.
> 
> Please see the behavior on an older MSAL to confirm.

* Since the input value `"{\"xms...\": ...}"` was in this case a string (regardless of whether it contains escaped quotation mark), the output is rightfully also the same string (with valid escaping).
* Out of curiosity, could the input be a json blob `{"xms_az_claim": ["GUID", "GUID2", "GUID3"]}`, and if so, will our implementation generate the output as the same json blob? [It probably will](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/4886/files#diff-f0d569d5c86679d53fd6a97ff8fa9030952e9511c986afd750b7c64e3263e506R77-R80), but I just want to call this out and ask.
